### PR TITLE
Scrabble Timed Exercise - Alex Barnes

### PR DESCRIPTION
**Pull request by abarnes26**
I really enjoyed this exercise.  I found it very challenging to have to go back and amend the existing code when new requirements were given.  The most challenging was the seven letter bonus.  I was very resistant to having such a bulky method for scoring points (if/elsif/elsif/else), but after working on it for some time I found it was the best way.  I nearly completed the last exercise, but ran out of time before I could get it to return just the key, and not a key/ value pair.

### CODE REVIEW

**Comment by abarnes26** [comment]
I figured out the last challenge just after time was up - 

```
  def highest_scoring_word(word_list)
    word_scores = {}
    word_list.map do |word|
      word_scores[word] = score(word)
     end
    final = word_scores.max_by do |word, score|
        score
     end
     final[0]
  end

```

File: test/scrabble_test.rb [hunk 1]
```diff
@@ -5,9 +5,73 @@
...
     assert_equal 4, Scrabble.new.score("f")
   end
+
+  def test_it_can_score_an_entire_word
+    assert_equal 8, Scrabble.new.score("dance")
+  end
+
+  def test_it_does_not_care_about_case
+    assert_equal 8, Scrabble.new.score("DaNcE")
+  end
```
Comment (AliSchlereth) [inline test/scrabble_test.rb]: Great job testing an edge case here.

File: test/scrabble_test.rb [hunk 2]
```diff
@@ -5,9 +5,73 @@
...
+  def test_it_can_score_empty_words
+    assert_equal 0, Scrabble.new.score("")
+    assert_equal 0, Scrabble.new.score(nil)
+  end
+
+  def test_it_can_multiply_score_for_a_single_letter
+    test_word = Scrabble.new.score_with_multipliers("d", [2])
+
+    assert_equal 4, test_word
+  end
```
Comment (AliSchlereth) [inline test/scrabble_test.rb]: I like how you are starting with the smallest possible scenario here.

File: test/scrabble_test.rb [hunk 3]
```diff
@@ -5,9 +5,73 @@
...
+    test_word = Scrabble.new.score_with_multipliers("down", [1,2,1,2], 3)
+
+    assert_equal 30, test_word
+  end
+
+  def test_it_can_add_a_seven_letter_bonus
+    test_word = Scrabble.new.score_with_multipliers("downtown", [1,1,1,2,1,1,1,2], 3)
+
+    assert_equal 81, test_word
+  end
```
Comment (AliSchlereth) [inline test/scrabble_test.rb]: I really like this series of 3 tests. Great job incrementally testing more and more. I would like to see the word multiplier be a different number between the two tests that cover it though.

File: lib/scrabble.rb [hunk 4]
```diff
@@ -15,4 +13,69 @@ def point_values
       "Y"=>4, "Z"=>10
     }
   end
+
+  def split_word(word)
+    if word != nil; word.upcase.chars
+    end
+  end
```
Comment (AliSchlereth) [inline lib/scrabble.rb]: You could do this in one line, something like this: 
`word.upcase.chars if word != nil`

File: lib/scrabble.rb [hunk 5]
```diff
@@ -15,4 +13,69 @@ def point_values
...
+
+  def score(word)
+    scored = 0
+    if word != nil; split_word(word).map do |letter|
+      scored += point_values[letter]
+      end
+    end
+    return scored
+  end
```
Comment (AliSchlereth) [inline lib/scrabble.rb]: I might re-organize these steps like this:
```
def score(word)
   return 0 if word == nil
  split_word(word).reduce(0) do |sum, letter|
     sum += letter
  end
end
``` 

That guard statement to start will make it so the rest of the method doesn't run if word is nil. 
Reduce is a good tool to use instead of .map + adding to a counter of sorts.

File: lib/scrabble.rb [hunk 6]
```diff
@@ -15,4 +13,69 @@ def point_values
...
+
+  def seven_letter_bonus(word)
+    if word.length >= 7
+      true
+    else
+     false
+    end
+  end
```
Comment (AliSchlereth) [inline lib/scrabble.rb]: Here you could have this method only run `word.length >= 7` since that phrase will return true of false

File: lib/scrabble.rb [hunk 7]
```diff
@@ -15,4 +13,69 @@ def point_values
...
+
+  def score_with_multipliers(word, letter_multiplier, word_mulitplier = nil)
+    scored = []
+    if word != nil; split_word(word).map do |letter|
+      scored << point_values[letter]
+      end
+    end
+    add_up_points(scored, letter_multiplier, word_mulitplier)
+  end
```
Comment (AliSchlereth) [inline lib/scrabble.rb]: I'd put a guard clause `return 0 if word == nil || word == ''` at this beginning of this method which catches both nil and an empty string. It will decrease some of your steps later on so you don't need to guard against nil in each method.

Comment (AliSchlereth) [reply lib/scrabble.rb]: I would default word_multiplier to 1 so that I can automatically multiply it times the total no matter what.

File: lib/scrabble.rb [hunk 8]
```diff
@@ -15,4 +13,69 @@ def point_values
...
+      score = (multiply_points(scored, letter_multiplier) + 10) * word_mulitplier
+    elsif word_mulitplier != nil && seven_letter_bonus(scored) == false
+      score = multiply_points(scored, letter_multiplier) * word_mulitplier
+    elsif word_mulitplier == nil && seven_letter_bonus(scored) == true
+      score = multiply_points(scored, letter_multiplier) + 10
+    else
+      score = multiply_points(scored, letter_multiplier)
+    end
+    return score
+  end
```
Comment (AliSchlereth) [inline lib/scrabble.rb]: If you have that guard clause in score_with_multipliers and default word_multiplier to 1, then you shouldn't need to put in so many branches to your if/else statement.

Comment (AliSchlereth) [reply lib/scrabble.rb]: Since each branch returns out a score, you don't need to save that to the variable score and then recall score. The branch will be the last thing run in the if/else statement and therefore the last thing to run in the method.

**Comment by AliSchlereth** [comment]
Generally great job at this challenge. I left some comments on how you could refactor things to decrease duplication and make things a little more streamlined.
